src/meson.build

```diff
@@ -0,0 +1,79 @@
+ffi_c_sources = [
+  'prep_cif.c',
+  'types.c',
+  'raw_api.c',
+  'java_raw_api.c',
+  'closures.c',
+  'tramp.c',
+]
+
+ffi_asm_sources = []
+
+if get_option('ffi-debug')
+  ffi_c_sources += ['debug.c']
+endif
+
+# Translate source files to source paths
+foreach f : c_sources
+  ffi_c_sources += targetdir / f
+endforeach
+foreach f : asm_sources
+  ffi_asm_sources += targetdir / f
+endforeach
+
+ffiinc = include_directories('..', '../include', targetdir)
+
+if is_msvc_like
+  # MSVC, cannot compile assembly files directly. They need to be preprocessed
+  # with the C compiler first, then compiled with MASM.
+  ffi_asm_sources = cc.preprocess(ffi_asm_sources,
+    output: '@BASENAME@.masm',
+    include_directories: ffiinc,
+  )
+  add_languages('masm', native: false)
+endif
+
+ffi_link_args = []
+ffi_map_file = []
+
+if get_option('ffi-build-versioned')
+  libffi_map_in = meson.project_source_root() / 'libffi.map.in'
+  arg_prefix = '-Wl,--version-script='
+  if cc.has_link_argument(arg_prefix + libffi_map_in)
```

Comment:
oh, that's unfortunate, there are tones of projects that use `has_link_argument` for this. We probably should do something in Meson, at least try to detect the situation and print a warning. Maybe even add an explicit `cc.has_version_script()`?

Reply:
Well, it's already on my to-do list to add direct support for

```meson
visibility_file: {
    'gnu': files('libffi.map'),
    'generic': 'libffi.syms',
    'msvc': custom_target(output: 'libffi.def', ...),
    'apple': 'libffi.syms.darwin'
}
```

Or something to that general effect.

Reply:
IMO using visibility files should be considered legacy

the right way to do things is to compile with `-fvisibility=hidden` by default, then mark specific symbols as visible *in the source*

Reply:
Fixed.

Reply:
@q66 I think it is only used if you want to set a version on symbols, not sure that can be done otherwise? Not sure it is even useful to have versioned symbols in libffi tbh.

Reply:
@q66,

That doesn't actually do what people want in all cases.
- ld version scripts are required if you want to add versions to the symbols, not just hide some -- you can use both visibility attributes and versioning together, even
- MSVC def files can apply to "just the shared library"
- if you're linking a shared library to a static library, you *might* want to hide external symbols regardless of their public-for-that-library visibility
- symbol visibility doesn't exist for fortran, but the linker doesn't have to care what the language supports

Reply:
it is possible with `__asm__` renaming on the function declarations

but i meant more like I'm not sure there is a point in adding the functionality to meson considering visibility should be handled like I said and as for versioned symbols, they are already rare and I'd consider them a bad idea for most part

Reply:
@eli-schwartz sure, and in those niche cases where that is necessary people can always do this manually like it's done right now

Reply:
as i see it, abstractions are only meaningful when they serve the purpose of erasing and hiding differences; in this case it seems to be more like a slightly more convenient but still fundamentally non-portable interface to functionality which is already niche and not used by most projects

Reply:
@q66 in any case, that's a discussion for a Meson issue/PR. As far as libffi is concerned, I think the problem is solved already.

Reply:
yup, agreed

Reply:
Yes, indeed, this is all a moot point. The only downside to implementing an "unneeded" option that is "legacy" is my wasted time spent implementing it, and I have a number of reasons why I regard this as a useful modern feature that isn't niche or legacy (some of which I mentioned here) and I'm happy with that logic, and so are people I've spoken to. :)